api/api.go

```diff
@@ -0,0 +1,127 @@
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/fluidkeys/api/v1structs"
+)
+
+const (
+	defaultBaseURL = "https://api.fluidkeys.com/v1/"
+	userAgent      = "fluidkeys"
+)
+
+// A Client manages communication with the Fluidkeys Server API.
+type Client struct {
+	client    *http.Client // HTTP client used to communicate with the API.
+	BaseURL   *url.URL     // Base URL for API requests
+	UserAgent string       // User agent used when communicating with the  API.
+}
+
+// NewClient returns a new Fluidkeys Server API client. If a nil httpClient is
+// provided, http.DefaultClient will be used.
+func NewClient() *Client {
+	baseURL, _ := url.Parse(defaultBaseURL)
+
+	return &Client{
+		client:    http.DefaultClient,
+		BaseURL:   baseURL,
+		UserAgent: userAgent,
+	}
+}
+
+// GetPublicKey attempts to get a single armorded public key.
+func (c *Client) GetPublicKey(email string) (string, *http.Response, error) {
+	url := fmt.Sprintf("email/%s/key", url.QueryEscape(email))
+	request, err := c.newRequest("GET", url, nil)
+	if err != nil {
+		return "", nil, err
+	}
+	decodedJSON := new(v1structs.GetPublicKeyResponse)
+	response, err := c.do(request, &decodedJSON)
+	if err != nil {
+		return "", response, err
+	}
+	return decodedJSON.ArmoredPublicKey, response, nil
+}
+
+// CreateSecret creates a secret for the given recipient
+func (c *Client) CreateSecret(recipientFingerprint string, armoredEncryptedSecret string) (*http.Response, error) {
+	sendSecretRequest := v1structs.SendSecretRequest{
+		RecipientFingerprint:   recipientFingerprint,
+		ArmoredEncryptedSecret: armoredEncryptedSecret,
+	}
+	url := fmt.Sprintf("secrets")
+	request, err := c.newRequest("POST", url, sendSecretRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := c.do(request, nil)
+	if err != nil {
+		return response, err
+	}
+
+	return response, nil
+}
+
+// newRequest creates an API request. A relative URL can be provided in urlStr,
```

Comment:
If i read the code correctly, it *only* accepts a relative path, right? If so, it could just be called `relativePath` and the docstring could say "relativePath is resolved relative to the BaseURL of the client"

Reply:
Actually I *didn't* read the code right.

The docstring tells me about using a relative path, hinting that I could instead use an absolute path (or URL?) but I'm not sure how.

Suggest: make it require a relative path, don't allow absolute path-or-url and make it simpler.